File: modules/swagger-jaxrs/src/test/scala/GenericsTest.scala
import io.swagger.jaxrs.Reader
import io.swagger.models._
import io.swagger.models.parameters.{BodyParameter, QueryParameter}
import io.swagger.models.properties._
import models.TestEnum
import org.junit.runner.RunWith
import org.scalatest.{FlatSpec, Matchers}
import org.scalatest.junit.JUnitRunner
import resources.ResourceWithGenerics

import scala.collection.JavaConverters.{asScalaBufferConverter, asScalaSetConverter}

@RunWith(classOf[JUnitRunner])
class GenericsTest extends FlatSpec with Matchers {
  val swagger = new Reader(new Swagger()).read(classOf[ResourceWithGenerics])
  val enumValues = (for (item <- TestEnum.values()) yield item.name()).toSet

  def testEnumCollection(p: QueryParameter, name: String) = {
    testCollection(p, name, "string", null)
    val schema = p.getItems.asInstanceOf[StringProperty]
    schema.getEnum.asScala.toSet should be(enumValues)
  }

  def testCollection(p: QueryParameter, name: String, `type`: String, format: String) = {
    p.getName should be(name)
    p.getType should be("array")
    p.getFormat should be(null)
    p.getCollectionFormat should be("multi")
    p.getItems should not be (null)
    val schema = p.getItems.asInstanceOf[Property]
    schema.getType should be(`type`)
    schema.getFormat should be(format)
  }

  def testScalar(p: QueryParameter, name: String, `type`: String, format: String) = {
    p.getName should be(name)
    p.getType should be(`type`)
    p.getFormat should be(format)
    p.getCollectionFormat should be(null)
    p.getItems should be(null)
  }

  def testGenericType(op: Operation, `type`: String) = {
    getBodyParameter(op, 0).getSchema.asInstanceOf[RefModel].getSimpleRef should be(`type`)
  }

  def getBodyParameter(op: Operation, index: Int) = {
    op.getParameters.get(index).asInstanceOf[BodyParameter]
  }

  def getOperation(name: String) = {
    swagger.getPath("/generics/" + name).getPost
  }

  def getQueryParameter(op: Operation, index: Int) = {
    op.getParameters.get(index).asInstanceOf[QueryParameter]
  }

  def getProperty(`type`: String, name: String) = {
    swagger.getDefinitions.get(`type`).asInstanceOf[ModelImpl].getProperties.get(name)
  }

  it should "check collections of integers" in {
    val op = getOperation("testIntegerContainers")
    op.getParameters().size() should be(8)
    testCollection(getQueryParameter(op, 0), "set", "integer", "int32")
    testCollection(getQueryParameter(op, 1), "list", "integer", "int32")
    testCollection(getQueryParameter(op, 2), "list2D", "string", null)
    testCollection(getQueryParameter(op, 3), "array", "integer", "int32")
    testCollection(getQueryParameter(op, 4), "arrayP", "integer", "int32")
    testScalar(getQueryParameter(op, 5), "scalar", "integer", "int32")
    testScalar(getQueryParameter(op, 6), "scalarP", "integer", "int32")
    testCollection(getQueryParameter(op, 7), "forced", "integer", "int32")
  }

  it should "check collections of strings" in {
    val op = getOperation("testStringContainers")
    op.getParameters().size() should be(5)
    val set = getQueryParameter(op, 0)
    testCollection(set, "set", "string", null)
    set.getItems.asInstanceOf[StringProperty].getEnum.asScala.toList should be(List("1", "2", "3"))
    testCollection(getQueryParameter(op, 1), "list", "string", null)
    testCollection(getQueryParameter(op, 2), "list2D", "string", null)
    testCollection(getQueryParameter(op, 3), "array", "string", null)
    testScalar(getQueryParameter(op, 4), "scalar", "string", null)
  }

  it should "check collections of objects" in {
    val op = getOperation("testObjectContainers")
    op.getParameters().size() should be(5)
    testCollection(getQueryParameter(op, 0), "set", "string", null)
    testCollection(getQueryParameter(op, 1), "list", "string", null)
    testCollection(getQueryParameter(op, 2), "list2D", "string", null)
    testCollection(getQueryParameter(op, 3), "array", "string", null)
    testScalar(getQueryParameter(op, 4), "scalar", "string", null)
  }

  it should "check collections of enumerations" in {
    val op = getOperation("testEnumContainers")
    op.getParameters().size() should be(5)
    testEnumCollection(getQueryParameter(op, 0), "set")
    testEnumCollection(getQueryParameter(op, 1), "list")
    testCollection(getQueryParameter(op, 2), "list2D", "string", null)
    testEnumCollection(getQueryParameter(op, 3), "array")
    val scalar = getQueryParameter(op, 4)
    testScalar(scalar, "scalar", "string", null)
    scalar.getEnum.asScala.toSet should be(enumValues)
  }

  it should "check collection of strings as body parameter" in {
    val op = getOperation("testStringsInBody")
    op.getParameters().size() should be(1)
    val p = getBodyParameter(op, 0)
    val strArray = p.getSchema.asInstanceOf[ArrayModel]
    strArray.getItems.asInstanceOf[Property].getType should be("string")
  }

  it should "check collection of objects as body parameter" in {
    val op = getOperation("testObjectsInBody")
    op.getParameters().size() should be(1)
    val p = getBodyParameter(op, 0)
    val objArray = p.getSchema.asInstanceOf[ArrayModel]
    objArray.getItems.asInstanceOf[RefProperty].getSimpleRef should be("Tag")
  }

  it should "check collection of enumerations as body parameter" in {
    val op = getOperation("testEnumsInBody")
    op.getParameters().size() should be(1)
    val p = getBodyParameter(op, 0)
    val enumArray = p.getSchema.asInstanceOf[ArrayModel]
    enumArray.getItems.asInstanceOf[StringProperty].getEnum.asScala.toSet should be(enumValues)
  }

  it should "check 2D array as body parameter" in {
    val op = getOperation("test2DInBody")
    op.getParameters().size() should be(1)
    val p = getBodyParameter(op, 0)
    val ddArray = p.getSchema.asInstanceOf[ArrayModel]
    ddArray.getItems.asInstanceOf[ArrayProperty].getItems.asInstanceOf[RefProperty].getSimpleRef should be("Tag")
  }

  it should "check parameters of generic types" in {
    val genericTypes = Set("GenericTypeString", "GenericTypeUUID", "GenericTypeGenericTypeString",
      "RenamedGenericTypeString", "RenamedGenericTypeRenamedGenericTypeString")
    (swagger.getDefinitions.keySet.asScala & genericTypes) should be(genericTypes)

    val opString = getOperation("testGenericType")
    testGenericType(opString, "GenericTypeString")
    val strValue = getProperty("GenericTypeString", "value")
    strValue should not be (null)
    strValue.getClass().getName() should be(classOf[StringProperty].getName)

    val opUUID = getOperation("testStringBasedGenericType")
    testGenericType(opUUID, "GenericTypeUUID")
    val uuidValue = getProperty("GenericTypeUUID", "value")
    uuidValue should not be (null)
    uuidValue.getClass().getName() should be(classOf[UUIDProperty].getName)

    val opComplex = getOperation("testComplexGenericType")
    testGenericType(opComplex, "GenericTypeGenericTypeString")
    val complexValue = getProperty("GenericTypeGenericTypeString", "value")
    complexValue should not be (null)
    complexValue.getClass().getName() should be(classOf[RefProperty].getName)
    complexValue.asInstanceOf[RefProperty].getSimpleRef should be("GenericTypeString")

    val opRenamed = getOperation("testRenamedGenericType")
    testGenericType(opRenamed, "RenamedGenericTypeRenamedGenericTypeString")
    val renamedComplexValue = getProperty("RenamedGenericTypeRenamedGenericTypeString", "value")
    renamedComplexValue should not be (null)
    renamedComplexValue.isInstanceOf[RefProperty] should be(true)
    renamedComplexValue.asInstanceOf[RefProperty].getSimpleRef should be("RenamedGenericTypeString")
  }

  it should "check generic result" in {
    val op = swagger.getPath("/generics/testGenericResult").getGet
    val schema = op.getResponses().get("200").getSchema
    schema.getClass.getName should be(classOf[RefProperty].getName)
    schema.asInstanceOf[RefProperty].getSimpleRef should be("GenericListWrapperTag")

    val entries = getProperty("GenericListWrapperTag", "entries")
    entries should not be (null)
    entries.getClass.getName should be(classOf[ArrayProperty].getName)

    val items = entries.asInstanceOf[ArrayProperty].getItems
    items.getClass.getName should be(classOf[RefProperty].getName)
    items.asInstanceOf[RefProperty].getSimpleRef should be("Tag")
  }
}

File: modules/swagger-jaxrs/src/test/java/io/swagger/GenericsTest.java
package io.swagger;

import com.google.common.base.Functions;
import com.google.common.collect.Collections2;
import com.google.common.collect.Sets;
import io.swagger.jaxrs.Reader;
import io.swagger.models.ArrayModel;
import io.swagger.models.Model;
import io.swagger.models.Operation;
import io.swagger.models.RefModel;
import io.swagger.models.Swagger;
import io.swagger.models.TestEnum;
import io.swagger.models.parameters.BodyParameter;
import io.swagger.models.parameters.QueryParameter;
import io.swagger.models.properties.ArrayProperty;
import io.swagger.models.properties.Property;
import io.swagger.models.properties.RefProperty;
import io.swagger.models.properties.StringProperty;
import io.swagger.models.properties.UUIDProperty;
import io.swagger.resources.ResourceWithGenerics;
import io.swagger.resources.generics.UserApiRoute;

import org.testng.annotations.Test;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertNotEquals;
import static org.testng.Assert.assertNotNull;
import static org.testng.Assert.assertTrue;

public class GenericsTest {
    private final Swagger swagger = new Reader(new Swagger()).read(ResourceWithGenerics.class);
    private final Set<String> enumValues = Sets.newHashSet(Collections2.transform(Arrays.asList(TestEnum.values()), Functions.toStringFunction()));

    private void testEnumCollection(QueryParameter p, String name) {
        testCollection(p, name, "string", null);
        StringProperty schema = (StringProperty) p.getItems();
        assertEquals(schema.getEnum(), enumValues);
    }

    private void testCollection(QueryParameter p, String name, String type, String format) {
        assertEquals(p.getName(), name);
        assertEquals(p.getType(), "array");
        assertEquals(p.getFormat(), null);
        assertEquals(p.getCollectionFormat(), "multi");
        assertNotEquals(p.getItems(), null);
        Property schema = p.getItems();
        assertEquals(schema.getType(), type);
        assertEquals(schema.getFormat(), format);
    }

    private void testScalar(QueryParameter p, String name, String type, String format) {
        assertEquals(p.getName(), name);
        assertEquals(p.getType(), type);
        assertEquals(p.getFormat(), format);
        assertEquals(p.getCollectionFormat(), null);
        assertEquals(p.getItems(), null);
    }

    private void testGenericType(Operation op, String type) {
        assertEquals(((RefModel) getBodyParameter(op, 0).getSchema()).getSimpleRef(), type);
    }

    private BodyParameter getBodyParameter(Operation op, int index) {
        return (BodyParameter) op.getParameters().get(index);
    }

    private Operation getOperation(String name) {
        return swagger.getPath("/generics/" + name).getPost();
    }

    private QueryParameter getQueryParameter(Operation op, int index) {
        return (QueryParameter) op.getParameters().get(index);
    }

    private Property getProperty(String type, String name) {
        return (swagger.getDefinitions().get(type)).getProperties().get(name);
    }

    @Test(description = "check collections of integers")
    public void checkCollectionsOfIntegers() {
        Operation op = getOperation("testIntegerContainers");
        assertEquals(op.getParameters().size(), 8);
        testCollection(getQueryParameter(op, 0), "set", "integer", "int32");
        testCollection(getQueryParameter(op, 1), "list", "integer", "int32");
        testCollection(getQueryParameter(op, 2), "list2D", "string", null);
        testCollection(getQueryParameter(op, 3), "array", "integer", "int32");
        testCollection(getQueryParameter(op, 4), "arrayP", "integer", "int32");
        testScalar(getQueryParameter(op, 5), "scalar", "integer", "int32");
        testScalar(getQueryParameter(op, 6), "scalarP", "integer", "int32");
        testCollection(getQueryParameter(op, 7), "forced", "integer", "int32");
    }

    @Test(description = "check collections of strings")
    public void checkCollectionsOfStrings() {
        Operation op = getOperation("testStringContainers");
        assertEquals(op.getParameters().size(), 5);
        QueryParameter set = getQueryParameter(op, 0);
        testCollection(set, "set", "string", null);
        assertEquals(((StringProperty) set.getItems()).getEnum(), Arrays.asList("1", "2", "3"));
        testCollection(getQueryParameter(op, 1), "list", "string", null);
        testCollection(getQueryParameter(op, 2), "list2D", "string", null);
        testCollection(getQueryParameter(op, 3), "array", "string", null);
        testScalar(getQueryParameter(op, 4), "scalar", "string", null);
    }

    @Test(description = "check collections of objects")
    public void checkCollectionsOfObjects() {
        Operation op = getOperation("testObjectContainers");
        assertEquals(op.getParameters().size(), 5);
        testCollection(getQueryParameter(op, 0), "set", "string", null);
        testCollection(getQueryParameter(op, 1), "list", "string", null);
        testCollection(getQueryParameter(op, 2), "list2D", "string", null);
        testCollection(getQueryParameter(op, 3), "array", "string", null);
        testScalar(getQueryParameter(op, 4), "scalar", "string", null);
    }

    @Test(description = "check collections of enumerations")
    public void checkCollectionsOfEnumerations() {
        Operation op = getOperation("testEnumContainers");
        assertEquals(op.getParameters().size(), 5);
        testEnumCollection(getQueryParameter(op, 0), "set");
        testEnumCollection(getQueryParameter(op, 1), "list");
        testCollection(getQueryParameter(op, 2), "list2D", "string", null);
        testEnumCollection(getQueryParameter(op, 3), "array");
        QueryParameter scalar = getQueryParameter(op, 4);
        testScalar(scalar, "scalar", "string", null);
        assertEquals(scalar.getEnum(), enumValues);
    }

    @Test(description = "check collection of strings as body parameter")
    public void checkCollectionsOfStringsAsBodyParameter() {
        Operation op = getOperation("testStringsInBody");
        assertEquals(op.getParameters().size(), 1);
        BodyParameter p = getBodyParameter(op, 0);
        ArrayModel strArray = (ArrayModel) p.getSchema();
        assertEquals(strArray.getItems().getType(), "string");
    }

    @Test(description = "check collection of objects as body parameter")
    public void checkCollectionsOfObjectsAsBodyParameter() {
        Operation op = getOperation("testObjectsInBody");
        assertEquals(op.getParameters().size(), 1);
        BodyParameter p = getBodyParameter(op, 0);
        ArrayModel objArray = (ArrayModel) p.getSchema();
        assertEquals(((RefProperty) objArray.getItems()).getSimpleRef(), "Tag");
    }

    @Test(description = "check collection of enumerations as body parameter")
    public void checkCollectionsOfEnumerationsAsBodyParameter() {
        Operation op = getOperation("testEnumsInBody");
        assertEquals(op.getParameters().size(), 1);
        BodyParameter p = getBodyParameter(op, 0);
        ArrayModel enumArray = (ArrayModel) p.getSchema();
        assertEquals(((StringProperty) enumArray.getItems()).getEnum(), enumValues);
    }

    @Test(description = "check 2D array as body parameter")
    public void check2DArrayAsBodyParameter() {
        Operation op = getOperation("test2DInBody");
        assertEquals(op.getParameters().size(), 1);
        BodyParameter p = getBodyParameter(op, 0);
        ArrayModel ddArray = (ArrayModel) p.getSchema();
        assertEquals(((RefProperty) ((ArrayProperty) ddArray.getItems()).getItems()).getSimpleRef(), "Tag");
    }

    @Test(description = "check parameters of generic types")
    public void checkParametersOfGenericTypes() {
        Set<String> genericTypes = new HashSet(Arrays.asList("GenericTypeString", "GenericTypeUUID", "GenericTypeGenericTypeString",
                "RenamedGenericTypeString", "RenamedGenericTypeRenamedGenericTypeString"));
        assertTrue(swagger.getDefinitions().keySet().containsAll(genericTypes));

        Operation opString = getOperation("testGenericType");
        testGenericType(opString, "GenericTypeString");
        Property strValue = getProperty("GenericTypeString", "value");
        assertNotEquals(strValue, null);
        assertEquals(strValue.getClass().getName(), StringProperty.class.getName());

        Operation opUUID = getOperation("testStringBasedGenericType");
        testGenericType(opUUID, "GenericTypeUUID");
        Property uuidValue = getProperty("GenericTypeUUID", "value");
        assertNotEquals(uuidValue, null);
        assertEquals(uuidValue.getClass().getName(), UUIDProperty.class.getName());

        Operation opComplex = getOperation("testComplexGenericType");
        testGenericType(opComplex, "GenericTypeGenericTypeString");
        Property complexValue = getProperty("GenericTypeGenericTypeString", "value");
        assertNotEquals(complexValue, null);
        assertEquals(complexValue.getClass().getName(), RefProperty.class.getName());
        assertEquals(((RefProperty) complexValue).getSimpleRef(), "GenericTypeString");

        Operation opRenamed = getOperation("testRenamedGenericType");
        testGenericType(opRenamed, "RenamedGenericTypeRenamedGenericTypeString");
        Property renamedComplexValue = getProperty("RenamedGenericTypeRenamedGenericTypeString", "value");
        assertNotEquals(renamedComplexValue, null);
        assertTrue(renamedComplexValue instanceof RefProperty);
        assertEquals(((RefProperty) renamedComplexValue).getSimpleRef(), "RenamedGenericTypeString");
    }

    @Test(description = "check generic result")
    public void checkGenericResult() {
        Operation op = swagger.getPath("/generics/testGenericResult").getGet();
        Property schema = op.getResponses().get("200").getSchema();
        assertEquals(schema.getClass().getName(), RefProperty.class.getName());
        assertEquals(((RefProperty) schema).getSimpleRef(), "GenericListWrapperTag");

        Property entries = getProperty("GenericListWrapperTag", "entries");
        assertNotEquals(entries, null);
        assertEquals(entries.getClass().getName(), ArrayProperty.class.getName());

        Property items = ((ArrayProperty) entries).getItems();
        assertEquals(items.getClass().getName(), RefProperty.class.getName());
        assertEquals(((RefProperty) items).getSimpleRef(), "Tag");
    }

    @Test(description = "scan model with Generic Type")
    public void scanModelWithGenericType() {
        final Swagger swagger = new Reader(new Swagger()).read(UserApiRoute.class);
        assertNotNull(swagger);
        final Model userEntity = swagger.getDefinitions().get("UserEntity");
        assertNotNull(userEntity);
        final Map<String, Property> properties = userEntity.getProperties();
        assertEquals(properties.size(), 2);
        assertNotNull(properties.get("id"));
        assertNotNull(properties.get("name"));
    }
}
